doc/sphinx/proof-engine/ltac.rst

```diff
@@ -904,11 +1043,38 @@ local definitions) with:
 
 .. tacn:: fun {+ @name } => @ltac_expr
 
-   Indeed, local definitions of functions are a syntactic sugar for binding
+   Indeed, local definitions of functions are syntactic sugar for binding
    a :n:`fun` tactic to an identifier.
 
    :tacn:`fun` is a :token:`binder_tactic`.
 
+The return value of a function is the value from the last expression.  For example:
```

Comment:
I think it's ambiguous what "last" means here.  It doesn't mean sequencing-last.

Reply:
What is a correct description then?

Reply:
I don't understand this sentence well enough to see what it's trying to get across.  I might say
```suggestion
The return value of a function is the value that results from evaluating its body.  For example:
```
(This assumes we've defined the body of a function, or that it'll be obvious.)  It's not clear to me what this sentence adds, though, and if I'm losing some bit of explanation here.

Reply:
I was thinking about this example you gave the other day,, where "ret." is the return value.  Is it like OCaml where the last @ltac_expr in a function is the return value?

```
Ltac cache_term val id :=
  let H := fresh in
  let id := fresh id in
  let ty := type of val in
  let __ := lazymatch goal with
            | _ => simple refine (let H : ty := _ in _);
                   [ transparent_abstract exact val using id | ]
            end in
  let ret := (eval cbv delta [H] in H) in
  let __ := lazymatch goal with _ => clear H end in
  ret.
```

Reply:
> Is it like OCaml where the last @ltac_expr in a function is the return value?

No.  In OCaml and Ltac2, `a; b` has the return value of `b`.  In Ltac1, `a; b` throws errors if either `a` or `b` has any return value other than `idtac`.  The only way to get a tactic to have a non-tactic return value is to ensure that it is a single expression with no sequencing which evaluates to something that is not a tactic.

Reply:
How about this?

A tactic can return a non-tactic value only by ending the expression with the idiom :n:`let … in @ident`.

Reply:
> A tactic can return a non-tactic value only by ending the expression with the idiom :n:`let … in @ident`.

That is worse; consider the following tactics which return non-tactic values:
```coq
Ltac unit := ().
Ltac get_goal := match goal with |- ?G => G end.
Ltac get_some_hyp := multimatch goal with H : _ |- _ => H end.
Ltac get_True := constr:(True).
Ltac head x := match x with ?f _ => head f | _ => x end.
Ltac thunk f := fun _ => f.
```
I don't believe there's a concise description of when things return tactics vs not.  Consider the following two tactics:
```coq
Ltac this_is_a_tactic_value :=
  let x := match constr:(True) with
           | True => idtac
           | False => constr:(I)
           end in
  x.
Ltac this_is_not_a_tactic_value :=
  let x := match constr:(False) with
           | True => idtac
           | False => constr:(I)
           end in
  x.
```
The only difference is which branch of the `match` is taken.

Reply:
I'll just say that functions can return values of any type and leave it at that.  The value types in Ltac seem a bit underdeveloped.  Maybe better in Ltac2.

Do you think it makes any sense to extend Ltac/Ltac2 to make them more like general-purpose scripting languages?  Not sure myself; fleshing out the language in a good way would be a lot of work and might distract developers from other more valuable tasks.

Reply:
Things are much better in Ltac2, which is a lot closer to OCaml in its semantics.  I don't think it makes much sense to flesh out Ltac1 that much, given that it'll eventually be deprecated in favor of Ltac2, I believe.